### ISSUE TITLE
Regex conditional groups are broken if no "else" expression is given

### ISSUE THREAD

**Issue by kendfrey** [dotnet/corefx#26787]
#### Scenario:

	Match match = Regex.Match("abc", "(?(?=x)bx)");
	Console.WriteLine("Success: {0}", match.Success);
	Console.WriteLine("Value: \"{0}\"", match.Value);
	Console.WriteLine("Index: {0}", match.Index);

#### Output:

	Success: True
	Value: ""
	Index: 1

#### Expected result:

I don't know. If leaving out the "else" expression isn't supported, it should throw an exception. Otherwise, matching anywhere seems the most reasonable option. In any case, if the condition fails, the result should not depend on the contents of the "if" expression.

#### Additional notes:

https://stackoverflow.com/q/48574288/785745

**Comment by danmosemsft** [comment]
Same in .NEt Framework.

**Comment by danmosemsft** [comment]
At first it seems fine to not have an "else". For example `Regex.Match("abc", "(?(?=bc)b)");` returns `b` and the same in [PCRE](https://regex101.com/r/okCsjM/1) and Perl.

In each case if the "else" is missing it seems to be equivalent to an empty pattern. ie., it always matches, but has no matching groups. In this case of course the assertion `(?=x)` is false so this is the result.

However if there is anything else in the pattern, that breaks down. Eg., although  `Regex.Match("abc", "(?(?=x)bx|a)bc")` matches `abc`,  `Regex.Match("abc", "(?(?=x)bx)abc")` fails to match at all.

As far as I can tell the behavior is the same as Perl. PCRE however does get `abc` using `(?(?=x)bx)abc`.

So I guess, always have an `else` otherwise you get odd behavior. If you are motivated, you could dig into the code to figure out "why" but I'm not sure this could be changed and still be confident of not breaking someone.

If you agree, we can close this.

cc @ViktorHofer

**Comment by kendfrey** [comment]
If this is intended, and there is some explanation/documentation of what the rationale behind the behaviour is, I'd be fine with closing it. It doesn't make sense to close it if it's a bug, though.

**Comment by ViktorHofer** [comment]
Thanks @kendfrey for the report. Thanks @danmosemsft for the thorough analysis. The documentation mentions that the else clause is optional: https://docs.microsoft.com/en-us/dotnet/standard/base-types/alternation-constructs-in-regular-expressions#Conditional_Expr

To make it absolutely clear we could add the phrasing that _if no else clause is provided and the condition doesn't match, the pattern matches with an empty capture_. I would also add the note that if something follows after the conditional match, no else clause is provided and the condition fails, the whole match will fail.

cc @rpetrusha @mairaw

**Comment by kendfrey** [comment]
I don't quite understand what you mean by:

> if something follows after the conditional match, no else clause is provided and the condition fails, the whole match fails.

Does this mean that `(?(?=x)a)a` would fail to match `"a"`? That doesn't seem right.

**Comment by ViktorHofer** [comment]
> Does this mean that (?(?=x)a)a would fail to match "a"? That doesn't seem right.

Correct.

```csharp
> var x = System.Text.RegularExpressions.Regex.Match("abc", "(?(?=x)bx)abc");
> x.Success
false
```

**Comment by kendfrey** [comment]
That doesn't make sense to me because that changes the behaviour of the conditional group based on context around it. Also, currently `(?(?=x)a)a` *does* match `"a"`.

**Comment by ViktorHofer** [comment]
You are right, something is wrong here:

```csharp
> System.Text.RegularExpressions.Regex.Match("abc", "(?(?=x)a)abc").Success
true
> System.Text.RegularExpressions.Regex.Match("abc", "(?(?=x)b)abc").Success
false
```

**Comment by danmosemsft** [comment]
It seems to me we don't fully understand what the engine is doing here enough to document it confidently. It would probably take someone stepping through the code. Either to figure out a fix (if it's not too breaking) or understand well enough to document it.

**Comment by ViktorHofer** [comment]
I agree. I will step through the code to get a better a understanding.

**Comment by kendfrey** [comment]
My working hypothesis is as described in the initial Stack Overflow question:

> It seems that a conditional group without an "else" expression will instead create a lookahead from the first character of the "if" expression and use that as the "else".

So `(?(?=abc)xyz)` is treated as `(?(?=abc)xyz|(?=x))`.

I don't know if this is accurate, but it does match everything I've seen so far.

I did take a look at the code, but I didn't understand it well enough to spot anything.

**Comment by ViktorHofer** [comment]
Tree parsing is correct:

```
Tree "(?(?=x)bx|a)bc":

Capture
 Concatenate
  Testgroup
   Require
    One (x)
   Multi (bx)
   One (a)
  Multi (bc)

Tree "(?(?=x)a)abc":

Capture
 Concatenate
  Testgroup
   Require
    One (x)
   One (a)
  Multi (abc)
```

**Comment by MohammadHamdyGhanem** [comment]
@ViktorHofer 
>You are right, something is wrong here:
> System.Text.RegularExpressions.Regex.Match("abc", "(?(?=x)a)abc").Success
true
> System.Text.RegularExpressions.Regex.Match("abc", "(?(?=x)b)abc").Success
false

But this will work:
```System.Text.RegularExpressions.Regex.Match("cabc", "(?(?=x)b)abc").Success // true```

note that ```(?(?=x)b)abc``` = ```"(?(x)b)abc```

And this will work (just add | ):

```System.Text.RegularExpressions.Regex.Match("cabc", "(?(?=x)b|)abc").Success // true```

So, if you look for quoick fix, you can just add | if the else part is messing!
But I advise to revise the Regex syntax and apply of my new suggestions as well :)

**Comment by MohammadHamdyGhanem** [comment]
After thinking of issues here and in https://github.com/dotnet/corefx/issues/30253
I think you should not substitute the missing else part with any thing, neither  empty string nor look ahead.
The right thing to do is using this algorithm:

```
regex = expr1(?(expr2)expr3)expr4

if (expr2 has a match)
    return (match expr1expr3expr4)

if (expr1="" and expr4 ="")
     return match failure.
else
    return (match expr1expr4)


```

**Comment by kendfrey** [comment]
@MohammadHamdyGhanem Having "" cause the match to fail is the opposite of how it normally works, which would be extremely confusing behaviour.

Otherwise, that looks proper (and is how I implemented it in my engine).

**Comment by MohammadHamdyGhanem** [comment]
@kendfrey 
> and is how I implemented it in my engine

Could you explain? I get confused. What "looks proper"?

Is there a link to your engine?
And can you implement any of these ideas?
https://github.com/dotnet/corefx/issues/29689
https://github.com/dotnet/corefx/issues/29937
I can use your engine in my Verex:
git remote add origin https://github.com/MohammadHamdyGhanem/Verex.git

**Comment by kendfrey** [comment]
@MohammadHamdyGhanem 

> Could you explain? I get confused. What "looks proper"?

Essentially just the Principle of Least Astonishment. It's what how I would expect it to behave.

> Is there a link to your engine?

It's [written in TypeScript](https://github.com/kendfrey/ncre) which may not be of much use to you. It's just intended to bring the features of .NET's regex to JS. (Still a WIP)
